src/app/binding/BindingTable.cpp

```diff
@@ -0,0 +1,237 @@
+{
+    if (device == nullptr)
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+    if (device->IsConnected())
+    {
+        return CHIP_ERROR_INCORRECT_STATE;
+    }
+    if (device->IsConnecting())
+    {
+        return CHIP_NO_ERROR;
+    }
+    Dnssd::Resolver::Instance().SetResolverDelegate(&mResolveDelegate);
+    CHIP_ERROR ret = Dnssd::Resolver::Instance().ResolveNodeId(device->GetPeerId(), Inet::IPAddressType::kAny);
+    return ret;
+}
+
+BindingTableEntry * BindingTable::FindBindingEntry(NodeId node, FabricIndex fabricIndex)
+{
+    return FindBindingEntry(BuildPeerId(node, fabricIndex));
+}
+
+BindingTableEntry * BindingTable::FindBindingEntry(PeerId peer)
+{
+    BindingTableEntry * foundEntry = nullptr;
+    mBindingEntries.ForEachActiveObject([&foundEntry, peer](BindingTableEntry * entry) {
+        if (entry->mPeerId == peer)
+        {
+            foundEntry = entry;
+            return false;
+        }
+        return true;
+    });
+    return foundEntry;
+}
+
+BindingTableEntry * BindingTable::FindBindingEntry(EndpointId endpoint, ClusterId cluster)
+{
+    BindingTableEntry * foundEntry = nullptr;
+    mBindingEntries.ForEachActiveObject([&foundEntry, endpoint, cluster](BindingTableEntry * entry) {
+        if (entry->mSourceEndpoint == endpoint && entry->mClusterId == cluster)
+        {
+            foundEntry = entry;
+            return false;
+        }
+        return true;
+    });
+    return foundEntry;
+}
+
+OperationalDeviceProxy * BindingTable::FindPeerDevice(NodeId node, FabricIndex fabricIndex)
+{
+    return FindPeerDevice(BuildPeerId(node, fabricIndex));
+}
+
+OperationalDeviceProxy * BindingTable::FindPeerDevice(PeerId peer)
+{
+    OperationalDeviceProxy * foundDevice = nullptr;
+    mPeerDevices.ForEachActiveObject([&foundDevice, peer](OperationalDeviceProxy * device) {
+        if (device->GetPeerId() == peer)
+        {
+            foundDevice = device;
+            return false;
+        }
+        return true;
+    });
+    return foundDevice;
+}
+
+void BindingTable::ResolveDelegate::OnNodeIdResolved(const Dnssd::ResolvedNodeData & nodeData)
+{
+    OperationalDeviceProxy * device = mBindingTable->FindPeerDevice(nodeData.mPeerId);
+    if (nodeData.mNumIPs > 0 && device != nullptr && !device->IsConnected() && !device->IsConnecting())
+    {
+        ChipLogProgress(AppServer, "BindingTable: NodeId resolved for 0xdevice %" PRIX64, nodeData.mPeerId.GetNodeId());
+        Transport::PeerAddress addr(nodeData.mAddress[0], Transport::Type::kUdp);
+        addr.SetInterface(nodeData.mInterfaceId);
+        device->UpdateDeviceData(addr, nodeData.mMrpRetryIntervalIdle, nodeData.mMrpRetryIntervalActive);
+        device->Connect(&mBindingTable->mOnDeviceConnected, &mBindingTable->mOnDeviceConnectionFailure);
+    }
+}
+
+void BindingTable::ResolveDelegate::OnNodeIdResolutionFailed(const PeerId & peerId, CHIP_ERROR error)
```

Comment:
There's an overall lack of any basic error handling here, especially in scenarios where the bound/peer device is not online (or reachable). In those cases, there needs to be some basic level of retrying happening.